wled00/espnow_broadcast.cpp

```diff
@@ -0,0 +1,249 @@
+static_assert(WLED_ESPNOW_MAX_MESSAGE_LENGTH <= ESP_NOW_MAX_DATA_LEN, "WLED_ESPNOW_MAX_MESSAGE_LENGTH must be <= 250 bytes");
+
+ESPNOWBroadcast espnowBroadcast;
+
+ESPNOWBroadcast::ESPNOWBroadcast() {    
+    _networkRxMessages = xRingbufferCreateNoSplit(sizeof(QueuedNetworkMessage), WLED_ESPNOW_MAX_QUEUED_MESSAGES);
+}
+
+bool ESPNOWBroadcast::setup() {
+
+    auto err = esp_event_loop_create_default();
+    if ( ERR_OK != err && ESP_ERR_INVALID_STATE != err ) {
+        return false;
+    }
+    err = esp_event_handler_instance_register(WIFI_EVENT, WIFI_EVENT_STA_START, onWiFiEvent, NULL, NULL);
+    if ( ERR_OK != err ) {
+        return false;
+    }
+    err = esp_event_handler_instance_register(WIFI_EVENT, WIFI_EVENT_STA_STOP, onWiFiEvent, NULL, NULL);
+    if ( ERR_OK != err ) {
+        return false;
+    }
+    err = esp_event_handler_instance_register(WIFI_EVENT, WIFI_EVENT_AP_START, onWiFiEvent, NULL, NULL);
+    if ( ERR_OK != err ) {
+        return false;
+    }
+
+    return setupWiFi();
+}
+
+bool ESPNOWBroadcast::setupWiFi() {
+    Serial.println("ESPNOWBroadcast::setupWiFi()");
+
+    _state.exchange(STOPPED);
+
+    // To enable ESPNow, we need to be in WIFI_STA mode
+    if ( !WiFi.mode(WIFI_STA) ) {
+        Serial.println("WiFi.mode() failed");
+        return false;
+    }
+    // and not have the WiFi connect
+    // Calling discount with tigger an async Wifi Event
+    if ( !WiFi.disconnect(false, true) ) {
+        Serial.println("WiFi.disconnect() failed");
+        return false;
+    }
+
+    return true;
+}
+
+
+void ESPNOWBroadcast::loop(size_t maxMessagesToProcess /*= 1*/) {
+    switch (espnowBroadcast._state.load()) {
+        case ESPNOWBroadcast::STARTING:
+            // if WiFI is in starting state, actually stat ESPNow from our main task thread.
+            start();
+            break;
+        case ESPNOWBroadcast::STARTED: {
+            auto ndx = maxMessagesToProcess;
+            while(ndx-- > 0) {
+                size_t size = 0;
+                auto *msg = (QueuedNetworkMessage*)xRingbufferReceive(_networkRxMessages, &size, 0);
+                if (msg) {
+                    if(size == sizeof(*msg)) {
+                        auto callback = _rxCallback;
+                        while( *callback ) {
+                            (*callback)(msg->mac, msg->data, msg->len);
+                            callback++;
+                        }
+                    }
+                    vRingbufferReturnItem(_networkRxMessages, (void *)msg);
+                } else {
+                    break;
+                }
+            }
+            break;
+        }
+        default:
+            break;
+    }
+}
+
+ESPNOWBroadcast::err_t ESPNOWBroadcast::send(const uint8_t* msg, size_t len) {
+    static const uint8_t broadcast[] = BROADCAST_ADDR_ARRAY_INITIALIZER;
+    return static_cast<ESPNOWBroadcast::err_t>(esp_now_send(broadcast, msg, len));
+}
+
+void ESPNOWBroadcast::start() {

```

Comment:
Having read other parts of the code, I wonder if you want to consolidate a few other places into this function
and instead of considering it a `start` function, instead think of it as a "make espnow match the state" function

Right now that logic lives in a combination of `loop`, `start`, and `onWifiEvent`

Instead you could move the `switch (espnowBroadcast._state.load()) {` into this function, and have handlers for each enum value:
* STARTED or STOPPED just returns
* STARTING does what you have here
* STOPPING (a new state) does the ` esp_err_t err = esp_now_deinit();` from onWifiEvent - which in its new form only changes the state.

That way, this method - on the main thread - can handle all retry timing and so on, plus it contains both the "set up" and "tear down" methods in one place.  All other threads are just changing the intended state or queueing received messages.